logstash-core/src/main/java/org/logstash/common/io/DeadLetterQueueWriter.java

```diff
@@ -220,6 +255,72 @@ private void innerWriteEntry(DLQEntry entry) throws IOException {
         lastWrite = Instant.now();
     }
 
+    private void executeAgeRetentionPolicy() throws IOException {
+        if (isOldestSegmentExpired()) {
+            deleteExpiredSegments();
+        }
+    }
+
+    private boolean isOldestSegmentExpired() {
+        if (retentionTime == null) {
+            return false;
+        }
+        final Instant now = clock.instant();
+        return oldestSegmentTimestamp
+                .map(t -> t.toInstant().isBefore(now.minus(retentionTime)))
+                .orElse(false);
+    }
+
+    private void deleteExpiredSegments() throws IOException {
+        // remove all the old segments that verifies the age retention condition
+        boolean cleanNextSegment;
+        do {
+            Path beheadedSegment = oldestSegmentPath.orElseThrow(() -> new IllegalStateException("DLQ writer can't find the oldest segment to drop on path: " + queuePath));
```

Comment:
Right, this should be resilient to `cleanConsumed` stealing segments. I'll fix it

Reply:
It's little more complicated. `deleteExpiredSegments` is executed only if `isOldestSegmentExpired` returns `true` . This means that the `oldestSegmentTimestamp` that we have fall out of the `retentionTime`.
The `oldestSegmentTimestamp` is valued in `updateOldestSegmentReference()` consistently with `oldestSegmentPath`.
So if we have a timestamp we have also a file. What could happen is that the optional `oldestSegmentPath` refers to a file which doesn't exists anymore (because of the `cleanConsumed` feature) and this is covered by https://github.com/elastic/logstash/pull/14255/files#diff-a0ee6ca8e72a830020520ea556f56e46ec1326e48593d9dfd1f252b70d3af45aR284-R287.

The  `oldestSegmentPath` 's orElseThrow 
```java
Path beheadedSegment = oldestSegmentPath.orElseThrow(() -> new IllegalStateException("DLQ writer can't find the oldest segment to drop on path: " + queuePath));
```
is a state consistency check on the writer side, which means:
"deleteExpiredSegments must be called with a value in oldestSegmentPath", doesn't mean that the value is valued, that's matter of the following lines
```java
try {
    Files.delete(beheadedSegment);
} catch (NoSuchFileException ex)  {}
```